std/range/package.d

```diff
@@ -8606,6 +8606,380 @@ if (isForwardRange!Source && hasLength!Source)
+form a new chunk.
+
+Params:
+    source = range from which the chunks will be selected
+    chunkSizes = range specifying the size of each chunk
+*/
+struct Chunks(Source, Sizes)
+if (isInputRange!Source && isInputRange!Sizes && is(ElementType!Sizes : size_t))
+{
+    static if (isForwardRange!Source && isForwardRange!Sizes)
+    {
+        /// Standard constructor
+        this(Source source, Sizes chunkSizes)
+        {
+            _source = source;
+            _chunkSizes = chunkSizes;
+            if (!_source.empty)
+                empty = false;
+            else
+                empty = true;
+        }
+
+        /// Input range primitives
+        public bool empty;
+
+        @property auto front()
+        {
+            assert(!empty, "Attempting to fetch the front of an empty range");
+
+            if (_chunkSizes.empty && !empty && !_source.empty)
+                return _source.save;
+
+            return _source.save.take(_chunkSizes.front());
+        }
+
+        void popFront()
+        {
+            assert(!empty, "Attempting to popFront an empty range");
+
+            // source range is longer than sum of sizes
+            if (_chunkSizes.empty && !_source.empty && !empty)
+            {
+                empty = true;
+                return;
+            }
+
+            _source.popFrontN(_chunkSizes.front());
+            _chunkSizes.popFront();
+
+            // length of the source range is smaller than or equal to the sum of sizes
+            if (_source.empty && !empty)
+                empty = true;
+        }
+
+        /// Forward range primitives
+        @property typeof(this) save()
+        {
+            return typeof(this)(_source, _chunkSizes);
+        }
+
+        static if (hasLength!Source)
+            @property size_t length()
+            {
+                Sizes sizesCopy = _chunkSizes.save();
+                size_t length = 0;
+                size_t sourceLength = _source.length;
+
+                while (!sizesCopy.empty && sourceLength > 0)
+                {
+                    import std.algorithm.comparison : min;
+                    sourceLength -= min(sourceLength, sizesCopy.front());
+                    sizesCopy.popFront();
+                    length++;
+                }
+                // source range is longer than the sum of sizes
+                if (sourceLength > 0)
+                    length++;
+
+                return length;
+            }
+
+        static if (isInfinite!Source && hasLength!Sizes)
+            @property size_t length()
+            {
+                return (_chunkSizes.length + 1);
+            }
+
+        static if (hasSlicing!Source && (isInfinite!Source || hasLength!(Source)))
+        {
+            auto opIndex(size_t index)
+            {
+                size_t start = 0;
+                Sizes sizesCopy = _chunkSizes.save();
+
+                while (index && !sizesCopy.empty)
```

Comment:
For opIndex and opSlice all of the first index lengths are needed to get the right slice from the source range so I can't see how this could become O(1). However I can change the behaviour a bit so length, back and popBack become faster: when source doesn't have enough elements, instead of ignoring the last given lengths an empty chunk will be formed. This way the range will have the same length as chunkSizes so instead of changing the source at popBack I can just remove the last length. Would that be alright? Or would it be better to just remove these operations and keep the current behaviour?

Reply:
I would (personally) remove these operations as it can have undesirable side-effects as it's easy to loose the overhead cost in generic algorithms.
However, as there are two sides to this argument and I would recommend to open a discussion in the general NG (aka Forum) and ask for opinions on this issue.
(Maybe there are other historic precedents, but IIRC so far everything in std.range does O(1) operations.)